### PR TITLE
fix(fetch): null values in body were not serialised

### PR DESCRIPTION
**Pull request by mxschmitt**
https://github.com/microsoft/playwright-dotnet/issues/2509

### CODE REVIEW

File: src/Playwright/Transport/Channels/APIRequestContextChannel.cs [hunk 1]
```diff
@@ -62,13 +63,15 @@ public APIRequestContextChannel(string guid, Connection connection, APIRequestCo
...
             ["params"] = parameters?.ToProtocol(),
             ["headers"] = headers?.ToProtocol(),
-            ["jsonData"] = jsonData,
             ["postData"] = postData != null ? Convert.ToBase64String(postData) : null,
             ["formData"] = formData?.ToProtocol(),
             ["multipartData"] = multipartData?.ToProtocol(),
-        };
-
-        var response = await Connection.SendMessageToServerAsync(Guid, "fetch", message).ConfigureAwait(false);
+        }.FilterNullValues();
```
Comment (campersau) [inline src/Playwright/Transport/Channels/APIRequestContextChannel.cs:69]: Isn't the null value filtering already done here?
https://github.com/microsoft/playwright-dotnet/blob/ffb7c21272c3f078c95976e7df60a41d9c53fbb7/src/Playwright/Transport/Connection.cs#L133-L138

Comment (campersau) [reply src/Playwright/Transport/Channels/APIRequestContextChannel.cs:69]: Ah you removed it, should have scrolled to the end before commenting 😆

File: src/Playwright/Helpers/DictionaryExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        return dictionary.Where(x => x.Value != null).ToDictionary(x => x.Key, x => FilterNullValuesInternal(x.Value)) as Dictionary<string, T>;
+    }
+
+    private static object FilterNullValuesInternal(object value)
+    {
+        if (value is ChannelBase)
+        {
+            return new Dictionary<string, object>() { { "guid", ((ChannelBase)value).Guid } };
+        }
```
Comment (campersau) [inline src/Playwright/Helpers/DictionaryExtensions.cs:46]:
```suggestion
        if (value is ChannelBase channelBase)
        {
            return new Dictionary<string, object>() { { "guid", channelBase.Guid } };
        }
```

File: src/Playwright/Helpers/DictionaryExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            return null;
+        }
+        if (value is string || value is int || value is bool || value is double || value is float || value is long || value is short || value is byte || value is uint || value is ulong || value is ushort || value is sbyte || value is decimal)
+        {
+            return value;
+        }
+        if (value is IDictionary dictionary)
+        {
+            return dictionary.Keys.Cast<string>().Where(key => dictionary[key] != null).ToDictionary(key => key, key => FilterNullValuesInternal(dictionary[key]));
```
Comment (campersau) [inline src/Playwright/Helpers/DictionaryExtensions.cs:63]: If you add a little helper like

```c#
private static IEnumerable<DictionaryEntry> EnumerateDictionaryEntries(IDictionary dict)
{
    foreach (DictionaryEntry item in dict)
    {
        yield return item;
    }
}
```

You can change it to this and avoid multiple indexes into the same key:

```suggestion
            return EnumerateDictionaryEntries(dictionary).Where(entry => entry.Value != null).ToDictionary(entry => entry.Key, entry => FilterNullValuesInternal(entry.Value));
```

**Comment by mxschmitt** [comment]
We'll go with https://github.com/microsoft/playwright-dotnet/pull/2532 instead.

**Pull request closed without merging**